File: py/sort/shuffle/KnuthDurstenfeld.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import random
from typing import List

class KnuthDurstenfeld:
    def shuffle(self, nums: List[int]) -> None:
        for i in range(len(nums) - 1, 0, -1):
            index = random.randrange(0, i + 1)
            nums[index], nums[i] = nums[i], nums[index]

File: java/sort/shuffle/KnuthDurstenfeld.java
package sort.shuffle;

import java.util.Random;

public class KnuthDurstenfeld {
    private static final Random _random = new Random();

    public void shuffle(int[] nums) {
        for (int i = nums.length - 1; i > 0; --i) {
            int index = _random.nextInt(i + 1);
            if (index != i) {
                nums[index] ^= nums[i];
                nums[i] ^= nums[index];
                nums[index] ^= nums[i];
            }
        }
    }
}
